### ISSUE TITLE
quantization deeplabv3(mobielentv2) error

### ISSUE THREAD

**Issue by raninbowlalala** [tensorflow/tensorflow#20867]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**:source
- **TensorFlow version (use command below)**:1.9.0
- **Python version**: 2.7.3
- **Bazel version (if compiling from source)**:0.12.0
- **GCC/Compiler version (if compiling from source)**:c++11
- **CUDA/cuDNN version**:7.5.18
- **GPU model and memory**:TITAN,12GB
- **Exact command to reproduce**:N/A

### Describe the problem
I want to train a quantization deeplabv3+(mobienetv2) model, use "mobilenetv2_coco_voc_trainaug" from https://github.com/tensorflow/models/blob/master/research/deeplab/g3doc/model_zoo.md.

### Source code / logs
 I add `tf.contrib.quantize.create_training_graph(quant_delay=0)` in line 315 https://github.com/tensorflow/models/blob/master/research/deeplab/train.py
but I got the error like below:
```
INFO:tensorflow:Training on train set
Traceback (most recent call last):
  File "deeplab/train.py", line 359, in <module>
    tf.app.run()
  File "/home/liufang/deeplab_venv/local/lib/python2.7/site-packages/tensorflow/python/platform/app.py", line 125, in run
    _sys.exit(main(argv))
  File "deeplab/train.py", line 281, in main
    tf.contrib.quantize.create_training_graph(quant_delay=0)
  File "/home/liufang/deeplab_venv/local/lib/python2.7/site-packages/tensorflow/contrib/quantize/python/quantize_graph.py", line 112, in create_training_graph
    freeze_bn_delay=freeze_bn_delay)
  File "/home/liufang/deeplab_venv/local/lib/python2.7/site-packages/tensorflow/contrib/quantize/python/quantize_graph.py", line 66, in _create_graph
    is_training=is_training)
  File "/home/liufang/deeplab_venv/local/lib/python2.7/site-packages/tensorflow/contrib/quantize/python/fold_batch_norms.py", line 54, in FoldBatchNorms
    graph, is_training, freeze_batch_norm_delay=freeze_batch_norm_delay)
  File "/home/liufang/deeplab_venv/local/lib/python2.7/site-packages/tensorflow/contrib/quantize/python/fold_batch_norms.py", line 100, in _FoldFusedBatchNorms
    fused_batch_norm=True))
  File "/home/liufang/deeplab_venv/local/lib/python2.7/site-packages/tensorflow/contrib/quantize/python/fold_batch_norms.py", line 323, in _ComputeBatchNormCorrections
    match.moving_variance_tensor + match.batch_epsilon)
TypeError: unsupported operand type(s) for +: 'NoneType' and 'float'
```

**Comment by raninbowlalala** [comment]
@drpngx Can you help me?

**Comment by drpngx** [comment]
Is `match.batch_epsilon` a `None`?

**Comment by raninbowlalala** [comment]
@drpngx I found "match.moving_variance_tensor", "match.moving_mean_tensor" and "match.bn_decay_mean_tensor" are None

**Comment by raninbowlalala** [comment]
@suharshs I find the code in tensorflow\contrib\quantize\python\fold_batch_norms.py, because the moving_mean_tensor, bn_decay_mean_tensor, moving_variance_tensor and bn_decay_var_tensor is not assign a value, can you take a look at it?
```
 # TODO(suharshs): Find a way to get rid of this inner match.
      for mul_match_result in moving_avg_mul_matcher.match_graph(graph):
        sub_op = mul_match_result.get_op(moving_average_sub_pattern)
        if sub_op.inputs[1].name == bn_op.outputs[1].name:
          # During training: Batch Mean is bn_op.outputs[1]
          moving_mean_tensor = sub_op.inputs[0]
          bn_decay_mean_tensor = mul_match_result.get_tensor(bn_decay_pattern)
        if sub_op.inputs[1].name == bn_op.outputs[2].name:
          # During training: Batch Var is bn_op.outputs[2]
          moving_variance_tensor = sub_op.inputs[0]
          bn_decay_var_tensor = mul_match_result.get_tensor(bn_decay_pattern)
```

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by G-mel** [comment]
@suharshs Anyway to resolve this error? I'm attempting to add quantization nodes into a loaded pre-trained keras model (based on Mobilenet) and I get this very same error. Keras seems to handle batch norm differently than TensorFlow, and thus the `moving_variance_tensor` and `moving_mean_tensor` are `None` and this error occurs.

**Comment by HimariO** [comment]
I had encountered the same error too. I resolve it by replacing `tf.layers.BatchNormalization` with `tf.contrib.slim.batch_norm`. So the arrangement of parameters in the graph matches what `_FindFusedBatchNorms` is looking for.

PS: `_FindFusedBatchNorms` is located in `tensorflow\contrib\quantize\python\fold_batch_norms.py`

**Comment by tensorflowbutler** [comment]
Nagging Assignees @suharshs, @raghuraman-k: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by raninbowlalala** [comment]
@suharshs But I saw tf.contrib.slim.batch_norm in  slim.mobilenetv2.py, I still meet this issue.

**Comment by saeed68gm** [comment]
@raninbowlalala  same here, @HimariO  where did you make this change?

**Comment by HimariO** [comment]
@saeed68gm I'm getting this error with my own project, but the issue is the same. You can look into `tensorflow\contrib\quantize\python\fold_batch_norms.py` to compare the graph structure `graph_matcher` is looking for and what `silm.mobilenetv2.py` created. I had taken a quick glance at `silm.mobilenetv2.py` 's graph but doesn't see anything wrong with it.

**Comment by saeed68gm** [comment]
@HimariO  Thank you for your response. The interesting thing is I was able to quantize deeplab using their training code (which uses the specialize implementation of mobilenetv2) but when I try to use the line tf.contrib.quantize.create_training_graph in mobilenetv2 implementation in slim, I get this error. So I guess I can compare the two and find out the difference....

**Comment by raninbowlalala** [comment]
@saeed68gm I will also check silm.mobilenetv2.py. If you have some progress, please comment here, we can discuss about this problem. Thank you very much!

**Comment by raninbowlalala** [comment]
@saeed68gm I find a solution, you can use "--fine_tune_batch_norm=true " instead of "--fine_tune_batch_norm=false". This error will  disappear.

**Comment by saeed68gm** [comment]
@raninbowlalala This is awesome, I will give it a try this afternoon! Thank a lot of sharing your results :)

**Comment by saeed68gm** [comment]
@raninbowlalala I am using a variation of mobiletnet v2 (imported from slim) and not sure why but I still get the error. The only thing that fine_tune_batch_norm seems to be doing is setting the is_training flag to True?

@HimariO I was not able to spot the difference between deeplab and mobilenetv2 from slim. Any luck for you?

**Comment by tensorflowbutler** [comment]
Nagging Assignees @suharshs, @raghuraman-k: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by saeed68gm** [comment]
@raninbowlalala @HimariO I debugged a lot and found out the source of the issue, but have no solution yet. The issue happens when you want to create an evaluation graph and you don't want to have "is_training=True" hence you won't have variables like moving variance. For right now, I just turn off eval graph when I want to quantize, but that's just a workaournd. Any idea what to do to have an eval graph and be able to quantize as well?

**Comment by tensorflowbutler** [comment]
Nagging Assignees @suharshs, @raghuraman-k: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by KapoorHitesh** [comment]
@saeed68gm how did you turn off eval? I am facing this issue to quantize faster rcnn model

**Comment by saeed68gm** [comment]
@KapoorHitesh it depends are your code. You gotta go dig into it and see where you create your eval graph and rip it all out.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by normandra** [comment]
Hey I'm having a similar problem. In my case I'm trying to do quantization aware training on the Cityscapes dataset. I tried adding the same line at 315, but I'm getting the following error instead: 

Any help would be very appreciated

```
Key MobilenetV2/Conv/act_quant/MobilenetV2/Conv/act_quant/max/biased not found in checkpoint
	 [[node save/RestoreV2 (defined at /home/mno/.local/lib/python3.6/site-packages/tensorflow/contrib/slim/python/slim/learning.py:659)  = RestoreV2[dtypes=[DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, ..., DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT], _device="/job:localhost/replica:0/task:0/device:CPU:0"](_arg_save/Const_0_0, save/RestoreV2/tensor_names, save/RestoreV2/shape_and_slices)]]

Caused by op 'save/RestoreV2', defined at:
  File "deeplab/train.py", line 403, in <module>
    tf.app.run()
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/platform/app.py", line 125, in run
    _sys.exit(main(argv))
  File "deeplab/train.py", line 396, in main
    save_interval_secs=FLAGS.save_interval_secs)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/contrib/slim/python/slim/learning.py", line 659, in train
    saver = saver or tf_saver.Saver()
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1102, in __init__
    self.build()
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1114, in build
    self._build(self._filename, build_save=True, build_restore=True)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1151, in _build
    build_save=build_save, build_restore=build_restore)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 795, in _build_internal
    restore_sequentially, reshape)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 406, in _AddRestoreOps
    restore_sequentially)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 862, in bulk_restore
    return io_ops.restore_v2(filename_tensor, names, slices, dtypes)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/ops/gen_io_ops.py", line 1466, in restore_v2
    shape_and_slices=shape_and_slices, dtypes=dtypes, name=name)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py", line 488, in new_func
    return func(*args, **kwargs)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3274, in create_op
    op_def=op_def)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1770, in __init__
    self._traceback = tf_stack.extract_stack()

NotFoundError (see above for traceback): Restoring from checkpoint failed. This is most likely due to a Variable name or other graph key that is missing from the checkpoint. Please ensure that you have not altered the graph expected based on the checkpoint. Original error:

Key MobilenetV2/Conv/act_quant/MobilenetV2/Conv/act_quant/max/biased not found in checkpoint
	 [[node save/RestoreV2 (defined at /home/mno/.local/lib/python3.6/site-packages/tensorflow/contrib/slim/python/slim/learning.py:659)  = RestoreV2[dtypes=[DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, ..., DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT], _device="/job:localhost/replica:0/task:0/device:CPU:0"](_arg_save/Const_0_0, save/RestoreV2/tensor_names, save/RestoreV2/shape_and_slices)]]

Traceback (most recent call last):
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1334, in _do_call
    return fn(*args)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1319, in _run_fn
    options, feed_dict, fetch_list, target_list, run_metadata)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1407, in _call_tf_sessionrun
    run_metadata)
tensorflow.python.framework.errors_impl.NotFoundError: Key MobilenetV2/Conv/act_quant/MobilenetV2/Conv/act_quant/max/biased not found in checkpoint
	 [[{{node save/RestoreV2}} = RestoreV2[dtypes=[DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, ..., DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT], _device="/job:localhost/replica:0/task:0/device:CPU:0"](_arg_save/Const_0_0, save/RestoreV2/tensor_names, save/RestoreV2/shape_and_slices)]]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1546, in restore
    {self.saver_def.filename_tensor_name: save_path})
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 929, in run
    run_metadata_ptr)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1152, in _run
    feed_dict_tensor, options, run_metadata)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1328, in _do_run
    run_metadata)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1348, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.NotFoundError: Key MobilenetV2/Conv/act_quant/MobilenetV2/Conv/act_quant/max/biased not found in checkpoint
	 [[node save/RestoreV2 (defined at /home/mno/.local/lib/python3.6/site-packages/tensorflow/contrib/slim/python/slim/learning.py:659)  = RestoreV2[dtypes=[DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, ..., DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT], _device="/job:localhost/replica:0/task:0/device:CPU:0"](_arg_save/Const_0_0, save/RestoreV2/tensor_names, save/RestoreV2/shape_and_slices)]]

Caused by op 'save/RestoreV2', defined at:
  File "deeplab/train.py", line 403, in <module>
    tf.app.run()
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/platform/app.py", line 125, in run
    _sys.exit(main(argv))
  File "deeplab/train.py", line 396, in main
    save_interval_secs=FLAGS.save_interval_secs)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/contrib/slim/python/slim/learning.py", line 659, in train
    saver = saver or tf_saver.Saver()
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1102, in __init__
    self.build()
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1114, in build
    self._build(self._filename, build_save=True, build_restore=True)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1151, in _build
    build_save=build_save, build_restore=build_restore)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 795, in _build_internal
    restore_sequentially, reshape)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 406, in _AddRestoreOps
    restore_sequentially)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 862, in bulk_restore
    return io_ops.restore_v2(filename_tensor, names, slices, dtypes)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/ops/gen_io_ops.py", line 1466, in restore_v2
    shape_and_slices=shape_and_slices, dtypes=dtypes, name=name)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py", line 488, in new_func
    return func(*args, **kwargs)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3274, in create_op
    op_def=op_def)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1770, in __init__
    self._traceback = tf_stack.extract_stack()

NotFoundError (see above for traceback): Key MobilenetV2/Conv/act_quant/MobilenetV2/Conv/act_quant/max/biased not found in checkpoint
	 [[node save/RestoreV2 (defined at /home/mno/.local/lib/python3.6/site-packages/tensorflow/contrib/slim/python/slim/learning.py:659)  = RestoreV2[dtypes=[DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, ..., DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT], _device="/job:localhost/replica:0/task:0/device:CPU:0"](_arg_save/Const_0_0, save/RestoreV2/tensor_names, save/RestoreV2/shape_and_slices)]]


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1556, in restore
    names_to_keys = object_graph_key_mapping(save_path)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1830, in object_graph_key_mapping
    checkpointable.OBJECT_GRAPH_PROTO_KEY)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/pywrap_tensorflow_internal.py", line 371, in get_tensor
    status)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/framework/errors_impl.py", line 528, in __exit__
    c_api.TF_GetCode(self.status.status))
tensorflow.python.framework.errors_impl.NotFoundError: Key _CHECKPOINTABLE_OBJECT_GRAPH not found in checkpoint

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "deeplab/train.py", line 403, in <module>
    tf.app.run()
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/platform/app.py", line 125, in run
    _sys.exit(main(argv))
  File "deeplab/train.py", line 396, in main
    save_interval_secs=FLAGS.save_interval_secs)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/contrib/slim/python/slim/learning.py", line 748, in train
    master, start_standard_services=False, config=session_config) as sess:
  File "/usr/lib/python3.6/contextlib.py", line 81, in __enter__
    return next(self.gen)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/supervisor.py", line 1004, in managed_session
    self.stop(close_summary_writer=close_summary_writer)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/supervisor.py", line 832, in stop
    ignore_live_threads=ignore_live_threads)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/coordinator.py", line 389, in join
    six.reraise(*self._exc_info_to_raise)
  File "/home/mno/.local/lib/python3.6/site-packages/six.py", line 693, in reraise
    raise value
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/supervisor.py", line 993, in managed_session
    start_standard_services=start_standard_services)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/supervisor.py", line 730, in prepare_or_wait_for_session
    init_feed_dict=self._init_feed_dict, init_fn=self._init_fn)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/session_manager.py", line 288, in prepare_session
    config=config)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/session_manager.py", line 218, in _restore_checkpoint
    saver.restore(sess, ckpt.model_checkpoint_path)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1562, in restore
    err, "a Variable name or other graph key that is missing")
tensorflow.python.framework.errors_impl.NotFoundError: Restoring from checkpoint failed. This is most likely due to a Variable name or other graph key that is missing from the checkpoint. Please ensure that you have not altered the graph expected based on the checkpoint. Original error:

Key MobilenetV2/Conv/act_quant/MobilenetV2/Conv/act_quant/max/biased not found in checkpoint
	 [[node save/RestoreV2 (defined at /home/mno/.local/lib/python3.6/site-packages/tensorflow/contrib/slim/python/slim/learning.py:659)  = RestoreV2[dtypes=[DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, ..., DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT], _device="/job:localhost/replica:0/task:0/device:CPU:0"](_arg_save/Const_0_0, save/RestoreV2/tensor_names, save/RestoreV2/shape_and_slices)]]

Caused by op 'save/RestoreV2', defined at:
  File "deeplab/train.py", line 403, in <module>
    tf.app.run()
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/platform/app.py", line 125, in run
    _sys.exit(main(argv))
  File "deeplab/train.py", line 396, in main
    save_interval_secs=FLAGS.save_interval_secs)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/contrib/slim/python/slim/learning.py", line 659, in train
    saver = saver or tf_saver.Saver()
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1102, in __init__
    self.build()
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1114, in build
    self._build(self._filename, build_save=True, build_restore=True)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1151, in _build
    build_save=build_save, build_restore=build_restore)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 795, in _build_internal
    restore_sequentially, reshape)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 406, in _AddRestoreOps
    restore_sequentially)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 862, in bulk_restore
    return io_ops.restore_v2(filename_tensor, names, slices, dtypes)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/ops/gen_io_ops.py", line 1466, in restore_v2
    shape_and_slices=shape_and_slices, dtypes=dtypes, name=name)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py", line 488, in new_func
    return func(*args, **kwargs)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3274, in create_op
    op_def=op_def)
  File "/home/mno/.local/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1770, in __init__
    self._traceback = tf_stack.extract_stack()

NotFoundError (see above for traceback): Restoring from checkpoint failed. This is most likely due to a Variable name or other graph key that is missing from the checkpoint. Please ensure that you have not altered the graph expected based on the checkpoint. Original error:

Key MobilenetV2/Conv/act_quant/MobilenetV2/Conv/act_quant/max/biased not found in checkpoint
	 [[node save/RestoreV2 (defined at /home/mno/.local/lib/python3.6/site-packages/tensorflow/contrib/slim/python/slim/learning.py:659)  = RestoreV2[dtypes=[DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, ..., DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT], _device="/job:localhost/replica:0/task:0/device:CPU:0"](_arg_save/Const_0_0, save/RestoreV2/tensor_names, save/RestoreV2/shape_and_slices)]]

```

**Comment by a00012025** [comment]
Hi, I also have the same issue here. I use the [MobileNet v2 Colab Example](https://colab.research.google.com/github/tensorflow/models/blob/master/research/slim/nets/mobilenet/mobilenet_example.ipynb) and try to quantize the model with `create_training_graph`, then I get the same error. Is there any suggestion on this issue? Thanks!

**Comment by Aspirinkb** [comment]
I trained my own  SSD model using `tensorflow/models/research/object_detection/` project and exported the inference graph by its `export_inference_graph.py` script, then I got the model.ckpt files and frozen_inference_graph.pb file.    
Now I want to quantize the trained model using quantization-aware training method, but I got the same error! The code  `match.moving_variance_tensor + match.batch_epsilon` arises error because `match.moving_variance_tensor` is NoneType, but why it is NoneType? and how fix it???

**Comment by Aspirinkb** [comment]
I have the same issue, but I fix it by using the method `tf.layers.batch_normalization( )` correctly.  
   
This error arises because `match.moving_variance_tensor` variable id `NoneType`, so you can not add it to another float number. But why this variable in graph's batch normalization layer is `NoneType`? I think you are using `tf.layers.batch_normalization( )` or some other batch normalization api in TensorFlow mistakenly.   

There is an argument named `training` in the [method](https://www.tensorflow.org/api_docs/python/tf/layers/batch_normalization). When you call this method, you must note that the `moving_mean` and `moving_variance` need to be updated when training, but not when inference. So the behavior of BN layer is different when in training mode or inference/test/eval mode, you just pass correct mode to the `tf.layers.batch_normalization(training=is_training )` method. If you do not pass correct argument to it, your model may not trained properly and `moving_mean` and `moving_variance` are not trained! So they are `NoneType` when quantizing the model.

**Comment by SanthoshRajendiran** [comment]
What are the changes to be done to enable quantization aware training for Deeplab V3 MobilenetV2??

**Comment by ajinkya933** [comment]
@SanthoshRajendiran did you find a solution to your earlier issue ?

> What and where are the changes to be done to enable quantization aware training for Deeplab V3 MobilenetV2??

**Comment by SanthoshRajendiran** [comment]
Not yet @ajinkya933 Just waiting for some helping hand.. Do let me know if you have figured out a solution for the same.

**Comment by anniezhi** [comment]
Did anyone solved this issue? I'm faced with exactly the same issue when adding a create_training_graph(). My model makes use of a pre-trained ResNet-v1 and is derived from it a bit. There is no explicit "batch normalization" in the derived model. Shall I find a quantized ResNet before further training?

**Comment by balazsbanto** [comment]
I was able to solve this issue by upgrading tensorflow version from 1.15 to 1.15.3.

**Comment by AMArostegui** [comment]
Same as @balazsbanto here. I was able to fix this by updating from 1.15.0 to 1.15.3

**Comment by mohantym** [comment]
Hi @raninbowlalala ! 1.x versions are not supported any more. It seems that it has been resolved in 1.15.3 from the above [comment ](https://github.com/tensorflow/tensorflow/issues/20867#issuecomment-716664588)though. Can we move this issue to closed status now?
